lua/nvim-treesitter/install.lua

```diff
@@ -302,100 +298,119 @@ local function install_lang(lang, ask_reinstall, cache_folder, install_folder, w
   }
 
   run_install(cache_folder, install_folder, lang, install_info, with_sync, generate_from_grammar)
+  return 0
 end
 
 local function install(options)
   options = options or {}
   local with_sync = options.with_sync
-  local ask_reinstall = options.ask_reinstall
   local generate_from_grammar = options.generate_from_grammar
   local exclude_configured_parsers = options.exclude_configured_parsers
 
-  return function (...)
-    if fn.executable('git') == 0 then
-      return api.nvim_err_writeln('Git is required on your system to run this command')
-    end
+  return function (bang)
+    return function (...)
+      if fn.executable('git') == 0 then
+        return api.nvim_err_writeln('Git is required on your system to run this command')
+      end
 
-    local cache_folder, err = utils.get_cache_dir()
-    if err then return api.nvim_err_writeln(err) end
+      local cache_folder, err = utils.get_cache_dir()
+      if err then return api.nvim_err_writeln(err) end
 
-    local install_folder, err = utils.get_parser_install_dir()
-    if err then return api.nvim_err_writeln(err) end
+      local install_folder, err = utils.get_parser_install_dir()
+      if err then return api.nvim_err_writeln(err) end
 
-    local languages
-    local ask
-    if ... == 'all' then
-      languages = parsers.available_parsers()
-      ask = false
-    elseif ... == 'maintained' then
-      languages = parsers.maintained_parsers()
-      ask = false
-    else
-      languages = vim.tbl_flatten({...})
-      ask = ask_reinstall
-    end
+      local languages
+      if ... == 'all' then
+        languages = parsers.available_parsers()
+      elseif ... == 'maintained' then
+        languages = parsers.maintained_parsers()
+      else
+        languages = vim.tbl_flatten({...})
+      end
 
-    if exclude_configured_parsers then
-      languages = utils.difference(languages, configs.get_ignored_parser_installs())
-    end
+      if exclude_configured_parsers then
+        languages = utils.difference(languages, configs.get_ignored_parser_installs())
+      end
 
-    if #languages > 1 then
-      reset_progress_counter()
-    end
+      if #languages > 1 then
+        reset_progress_counter()
+      end
 
-    for _, lang in ipairs(languages) do
-      install_lang(lang, ask, cache_folder, install_folder, with_sync, generate_from_grammar)
+      local skipped = 0
+      for _, lang in ipairs(languages) do
+        skipped = skipped + install_lang(bang, lang, cache_folder, install_folder, with_sync, generate_from_grammar)
+      end
+      if skipped > 0 then
+        print('Parsers already installed. Use :TSInstall! to force reinstallation.')
```

Comment:
I think we should print this message on each parser that is already installed, or print the whole list here, otherwise isn't clear which parsers are already installed. Also, kind of having the input asking for re-installation was nice, maybe we can leave it since now is possible to avoid the prompt with the bang? We could make the `ask_reinstall` a global option.

Reply:
Listing them all becomes quite verbose when they are all installed and up to date. Imagine the scenario you have a post update hook in :PlugUpdate which calls :TSUpdate - it can be quite a lot of noise. We could concatenate a single-line list, but imho people are mainly interested in the ones that got updated, not the ones that were ignored.

And certainly don't want to prompt whether to overwrite for every parser in that scenario. Simply running :PlugUpdate causes the user to have to answer ~30 or so y/n prompts that they have no real way of giving a proper answer and just wanted to get things updated.

I don't think the installation/updates should be interactive in any scenario. It's better for scripting/deployment/installation, more consistent with other commands and is a better user experience.

imho there's just a few scenarios

- initial install of parsers, non-interactive
- update/sync of existing parsers, non-interactive
- clean re-install of parsers, non-interactive

I can't imagine anyone actually setting the 'ask_reinstall' option - seems like a nvim-treesitter developer option only. Although potentially the :TSInstallSync or :TSUpdateSync could be non-interactive while the :TSInstall / :TSUpdate could be interactive.

Note the :TSInstall and :TSIntsallSync commands are fairly redundant now given :TSUpdate will install if its missing. The only benefit for :TSInstall! over :TSUpdate is to allow a forced re-install via the bang attribute.

Reply:
Can you change this to `skipped..'parsers already installed'`?

Reply:
I'm also rather in favor of the current behavior asking reinstall. Non-interactive workflows can always use the lua functions directly.